File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    static int[][] dp;
    static int N;


    static void triangle() throws Exception {
        int i, j;
        int max = Integer.MIN_VALUE;
        for (i = 1; i <= N; i++) {
            for (j = 1; j <= i; j++) {
                if (j == 1) {
                    dp[i][j] += dp[i - 1][j];
                } else if (i == j) {
                    dp[i][j] += dp[i - 1][j - 1];
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - 1]) + dp[i][j];
                }
                if (max < dp[i][j]) max = dp[i][j];
            }
        }
        bw.write(max + "\n");
    }

    public static void main(String[] args) throws Exception {
        N = Integer.parseInt(br.readLine());
        dp = new int[N + 1][N + 1];
        for (int i = 1; i <= N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 1; j <= i; j++) {
                dp[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        triangle();
        bw.close();
    }
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm> 

using namespace std;

int solution(vector<vector<int>> triangle) {
    int n = triangle.size();
    
    vector<vector<int>> dp = triangle;
    
    for(int i = 1; i < n; i++) { 
        for(int j = 0; j <= i; j++) {
            if(j == 0) {
                dp[i][j] += dp[i-1][j];
            }
            else if(j == i) {
                dp[i][j] += dp[i-1][j-1];
            }
            else {
                dp[i][j] += max(dp[i-1][j-1], dp[i-1][j]);
            }
        }
    }
    
    return *max_element(dp[n-1].begin(), dp[n-1].end());
}